_posts/2020-08-26-gsoc-webgpu.md

```diff
@@ -0,0 +1,89 @@
+- Integrate WebGPU with WebRender for rendering the textures on HTML canvas.
+- Setup proper model model for async error recording.
+
+The final goal was to be able to run the live examples at https://austineng.github.io/webgpu-samples/ and pass a fair amount of the CTS.
+
+##Implementation
+Setting up a proper Id rotation for the GPU resources was our first priority. Since it was agreed to have a single wgpu thread for each content process, I had to ensure that each Id generated were unique. This meant sharing the Identity Hub among all threads via `Arc` and `Mutex`. For recycling the Ids, wgpu exposes an `IdentityHandler` trait that must be implemented on the server-side interface of the browser and wgpu. This facilitates the following- when wgpu detects that an object has been dropped by the user (which is some time after the actual drop/garbage collection), wgpu calls the trait methods that are responsible for releasing the Id. In our case they send a message to the content process to free the Id and make it available for reuse.
+
+Implementing the DOM Interfaces was pretty straight forward. A DOM object is just an opaque handle to an actual GPU resource. Whenever a method, that performs an operation, is called on a DOM object there are 2 things to be done- convert the IDL types to wgpu types and send a message to the server to perform the operation. Most of the validation is done on in wgpu itself.
+
+##Presentation
+WebGPU textures can be rendered to HTML canvas via `GPUCanvasContext`, which can be obtained from `canvas.getContext('gpupresent')`. All rendered images are served to WebRender as `ExternalImages` for rendering purposes. This is done via an async software presentation path. Each new `GPUCanvasContext` object is assigned a new `ExternalImageId` and a new swap chain is assigned a new `ImageKey`. Since WebGPU threads are spawned on-demand, an image handler for WebGPU is initialized at startup, stored in `Constellation`, and supplied to threads at the time of spawn. Each time `GPUSwapChain.getCurrentTexture()` is called the canvas is marked as dirty which is then flushed at the time of reflow. At the time of flush, a message is sent to the wgpu server to update the image data provided to WebRender. The following happens after this-
+- The contents of the rendered texture are copied to a buffer.
+- Buffer is mapped asynchronously for read.
+- The data read from the buffer is copied to a staging area in `PresentionData`. `PresentationData` stores the data and all the required machinery for this async presentation belt.
+
+https://github.com/servo/servo/blob/669b16f2c054bd038b7a3c69985076607e140b7f/components/webgpu/lib.rs#L1353-L1365
+
+- When WebRender wants to read the data, it locks on the data to prevent it from being altered during read. Data is served in the form of raw bytes.
+
+The above process is not the best one but the only option available to us for now. This also causes a few empty frames to be rendered at the start.
```

Comment:
would be nice to add another positive note: it works everywhere, and it's a great fallback path while we'll be adding hardware accelerate presentation in the future.